# Computer Infrastructure Tasks Summary
## Task 1
A directory called "data" was created using the `mkdir` command, and two subdirectories - "timestamps" and "weather" were created inside this. The `cd` command was used to navigate into the data directory in order to make the subdirectories. 

```bash
mkdir data
cd data
mkdir timestamps
mkdir weather
```

## Task 2
Inside the timestamps directory, the `date` command was used to output the current timestamp. A file was created called `now.txt` using the `touch` command, and the `>>` operator was used to append the output of `date` to the file. This was repeated, and the contents of the file were inspected with the `more` command (this can also be accomplished with `cat`, which prints the contents of the file to the terminal)

```bash
cd timestamps
date
touch now.txt
date >> now.txt
date >> now.txt
... etc.
more now.txt
```

## Task 3
A new file called `formatted.txt` was created with the `touch` command, and the current timestamp was appended to it using the requested format (`YYYYmmdd_HHMMSS`) which was supplied as an argument to `date`. The `man` command was used to see how this argument should be structured - using `%` indicators for each timestamp component, eg. `%Y` for the year, `%m` for the month, etc.

```bash
touch formatted.txt
man date
date +"%Y%m%d_%H%M%S" >> formatted.txt
```

## Task 4

The above date format was used with the `touch` command in order to create a file with the formatted timestamp as part of its filename. Backticks were used rather than `>>`.

```bash
touch `date +"%Y%m%d_%H%M%S"`.txt
```

## Task 5
In the `date/weather` directory, `wget` was used to retrieve the json weather content from the url: `https://prodapi.metweb.ie/observations/athenry/today`

The `-O` flag was used to specify an output file - `weather.json` 

```bash
cd ../weather
wget -O weather.json https://prodapi.metweb.ie/observations/athenry/today
```

Note:
`wget` is not available by default on OSX, so the `brew` package manager was installed and (added to the `$PATH` so it could be called from a terminal session), which was then used to install `wget`.

```bash
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
echo >> ~/.zprofile
echo 'eval "$(/opt/homebrew/bin/brew shellenv)"' >> ~/.zprofile
eval "$(/opt/homebrew/bin/brew shellenv)"

brew install wget
```

## Task 6
Task 4 and task 5 were combined download the weather json data and save it to a timestamp formatted filename.

```bash
wget -O `date +"%Y%m%d_%H%M%S"`.json https://prodapi.metweb.ie/observations/athenry/today
```

## Task 7
The command from task 6 was added to a file called `weather.sh`. Its permissions were changed with `chmod` in order to make it executable (`+x`). This required the `sudo` command to run `chmod` as the superuser (root), which has the necessary permissions to make this change. A bash shebang (`#!bash`) was also added to the top of the file, so that the caller knows what language the script is to run with.

```bash
touch weather.sh
sudo chmod +x weather.sh
```

```bash
# weather.sh
#!bash
wget -O data/weather/`date +"%Y%m%d_%H%M%S"`.json https://prodapi.metweb.ie/observations/athenry/today
```

In [1]:
import pandas as pd
df = pd.read_json("data/weather/20241212_155354.json")
df.head()

name  temperature symbol weatherDescription      text  windSpeed  \
0  Athenry            4    04n             Cloudy  "Cloudy"          7   
1  Athenry            4    04n             Cloudy  "Cloudy"          6   
2  Athenry            4    04n             Cloudy  "Cloudy"          6   
3  Athenry            4    04n             Cloudy  "Cloudy"          7   
4  Athenry            4    04n             Cloudy  "Cloudy"          4   

  windGust cardinalWindDirection  windDirection  humidity  rainfall  pressure  \
0        -                     E             90        78         0      1035   
1        -                    SE            135        79         0      1034   
2        -                    SE            135        79         0      1034   
3        -                    SE            135        75         0      1034   
4        -                    SE            135        80         0      1034   

    dayName       date reportTime  
0  Thursday 2024-12-12      00:00  
1  Thursday 2024-12-12      01:00  
2  Thursday 2024-12-12      02:00  
3  Thursday 2024-12-12      03:00  
4  Thursday 2024-12-12      04:00

In [9]:
df['windGust']

0     -
1     -
2     -
3     -
4     -
5     -
6     -
7     -
8     -
9     -
10    -
11    -
12    -
13    -
14    -
15    -
Name: windGust, dtype: object

In [2]:
len(df)

16

In [3]:
list(df)

['name',
 'temperature',
 'symbol',
 'weatherDescription',
 'text',
 'windSpeed',
 'windGust',
 'cardinalWindDirection',
 'windDirection',
 'humidity',
 'rainfall',
 'pressure',
 'dayName',
 'date',
 'reportTime']

In [8]:
len(list(df))

15

In [4]:
df.dtypes

name                             object
temperature                       int64
symbol                           object
weatherDescription               object
text                             object
windSpeed                         int64
windGust                         object
cardinalWindDirection            object
windDirection                     int64
humidity                          int64
rainfall                          int64
pressure                          int64
dayName                          object
date                     datetime64[ns]
reportTime                       object
dtype: object

In [7]:
df.describe()

temperature  windSpeed  windDirection   humidity  rainfall  \
count    16.000000  16.000000      16.000000  16.000000      16.0   
mean      3.875000   6.125000     109.687500  79.125000       0.0   
min       3.000000   4.000000      90.000000  75.000000       0.0   
25%       3.750000   4.000000      90.000000  78.000000       0.0   
50%       4.000000   6.500000      90.000000  79.000000       0.0   
75%       4.000000   7.000000     135.000000  80.000000       0.0   
max       5.000000   9.000000     135.000000  83.000000       0.0   
std       0.619139   1.707825      23.055639   2.028957       0.0   

          pressure                 date  
count    16.000000                   16  
mean   1033.187500  2024-12-12 00:00:00  
min    1031.000000  2024-12-12 00:00:00  
25%    1033.000000  2024-12-12 00:00:00  
50%    1033.000000  2024-12-12 00:00:00  
75%    1034.000000  2024-12-12 00:00:00  
max    1035.000000  2024-12-12 00:00:00  
std       1.108678                  NaN

The weather data contains 16 rows (as it was nearly 4pm at the time of retrieval) across 15 columns.

- **name**: The name of the weather station - in this case Athenry
- **temperature**: The dry bulb temperature in C, stored as an integer with values between 3 and 5 degrees.
- **symbol**: The weather icon, 04n corresponds to cloudy, as seen below.
- **weatherDescription**: A description of the weather (Cloudy)
- **text**: A description of the weather (Cloudy)
- **windSpeed**: The wind speed in knots, between 6 and 9 with a mean of 6.125 in this instance.
- **windGust**: The gust wind speed presumably, this is missing in this file (all rows contain "-") and not referred to on data.gov.ie
- **cardinalWindDirection**: The direction of the wind expressed in compass abbreviations, E=East / SE=South-East for example
- **windDirection**: The wind direction in degrees, 0 corresponding to north, 90 corresponding to east, 180 corresponding to south, etc.
- **humidity**: the humidity %, stored as an integer - between 75% and 83% on this day.
- **rainfall**: The rainfall in mm, also an integer - no rainfall was reported in this file
- **pressure**: The pressure in mbar, the average pressure in this case was 1033mbar
- **dayName**: The day name, Thursday in this case
- **date**: The date - 12/December/2024, stored in %Y-%m-%d format
- **reportTime**: The time of the report, each row corresponds to an hour of the day eg. 00:00, 10:00, etc

*Weather symbol 04n:*

<img src="https://openweathermap.org/img/wn/04n@2x.png"/>